primitives/npos-elections/src/pjr.rs

```diff
@@ -0,0 +1,523 @@
+/// This could cause the caller runtime to have to read the entire list of voters, which is assumed
+/// to be expensive.
+///
+/// A sensible user of this module should make sure that the PJR check is executed and checked as
+/// little as possible, and take sufficient economical measures to ensure that this function cannot
+/// be abused.
+fn prepare_pjr_input<AccountId: IdentifierT>(
+	supports: &Supports<AccountId>,
+	all_candidates: Vec<AccountId>,
+	all_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+) -> (Vec<CandidatePtr<AccountId>>, Vec<Voter<AccountId>>) {
+	// collect all candidates and winners into a unified `Vec<Candidate>`.
+	let mut candidates_index: BTreeMap<AccountId, usize> = BTreeMap::new();
+
+	// dump the staked assignments in a voter-major map for faster access down the road.
+	let mut assignment_map: BTreeMap<AccountId, Vec<(AccountId, ExtendedBalance)>> = BTreeMap::new();
+	for (winner_id, Support { voters, .. }) in supports.iter() {
+		for (voter_id, support) in voters.iter() {
+			assignment_map.entry(voter_id.clone()).or_default().push((winner_id.clone(), *support));
+		}
+	}
+
+	// Convert Suppports into a SupportMap
+	//
+	// As a flat list, we're limited to linear search. That gives the production of `candidates`,
+	// below, a complexity of `O(s*c)`, where `s == supports.len()` and `c == all_candidates.len()`.
+	// For large lists, that's pretty bad.
+	//
+	// A `SupportMap`, as a `BTreeMap`, has access timing of `O(lg n)`. This means that constructing
+	// the map and then indexing from it gives us timing of `O((s + c) * lg(s))`. If in the future
+	// we get access to a deterministic `HashMap`, we can further improve that to `O(s+c)`.
+	//
+	// However, it does mean allocating sufficient space to store all the data again.
+	let supports: SupportMap<AccountId> = supports.iter().cloned().collect();
+
+	let candidates = all_candidates.into_iter().enumerate().map(|(i, c)| {
+		candidates_index.insert(c.clone(), i);
+
+		// set the backing value and elected flag if the candidate is among the winners.
+		let who = c;
+		let maybe_support = supports.get(&who);
+		let elected = maybe_support.is_some();
+		let backed_stake = maybe_support.map(|support| support.total).unwrap_or_default();
+
+		Candidate { who, elected, backed_stake, ..Default::default() }.to_ptr()
+	}).collect::<Vec<_>>();
+
+	// collect all voters into a unified Vec<Voters>.
+	let voters = all_voters.into_iter().map(|(v, w, ts)| {
+		let mut edges: Vec<Edge<AccountId>> = Vec::with_capacity(ts.len());
+		for t in ts {
+			if edges.iter().any(|e| e.who == t) {
+				// duplicate edge.
+				continue;
+			}
+
+			if let Some(idx) = candidates_index.get(&t) {
+				// if this edge is among the assignments, set the weight as well.
+				let weight = assignment_map
+					.get(&v)
+					.and_then(|d| d.iter().find_map(|(x, y)| if x == &t { Some(y) } else { None }))
+					.cloned()
+					.unwrap_or_default();
```

Comment:
I wonder if this can ever happen? No afaik, but we're being defensive here and it is okay.